DOC: Clarify documentation of 'ambiguous' parameter

bartaelterman opened the pull request:
when first reading the docs, the `ambiguous` parameter was not clear to me. I propose to add a line of documentation and include an example that exactly indicates the issue that the `ambiguous` parameter solves.

bartaelterman committed:
Add documentation line with example for the ambiguous parameter of tz_locaclize

pep8speaks commented:
Hello @bartaelterman! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/generic.py`](https://github.com/bartaelterman/pandas/blob/f25ed6b9d3531a0a9d20308123b7b12839542c9c/pandas/core/generic.py) !

mroeschke commented on pandas/core/generic.py:
```diff
@@ -8623,6 +8623,13 @@ def tz_localize(self, tz, axis=0, level=None, copy=True,
             Also make a copy of the underlying data
         ambiguous : 'infer', bool-ndarray, 'NaT', default 'raise'
+            When clocks moved backward due to DST, ambiguous times may arise.
+            E.g. in Central European Time (UTC+01), when going from 03:00 DST
```
E.g. --> For example

mroeschke commented on pandas/core/generic.py:
```diff
@@ -8623,6 +8623,13 @@ def tz_localize(self, tz, axis=0, level=None, copy=True,
+            E.g. in Central European Time (UTC+01), when going from 03:00 DST
+            to 02:00 non-DST, the naive time 02:30:00 occurs both at
+            00:30:00 UTC and at 01:30:00 UTC. In such a situation, the
+            `ambiguous` parameter can be used to correctly localize these
```
I'd remove this last sentence. There's technically no _correct_ way to localize ambiguous times, pandas just provides the options below.

bartaelterman replied:
I would actually argue that that is exactly what the `ambiguous` parameter does. There is no way to guess it correctly, that's for sure (that's what `ambiguous='infer'` would do). But if you have, say, 5 records of which you know the first 2 should be interpreted as DST and the other 3 as non-DST, you can use `ambiguous=np.array([True, True, False, False, False])` to do exactly that.
We encountered a situation where we had exactly this. There was a separate column indicating some sort of time zone ('CET' / 'CEST') but since 'CEST' is not an official timezone, we could not use that column as the timezone name to localize the times. However, we could use `ambiguous=df['timezone_col']=='CEST'`. In that case I think we were able to _correctly_ localize the times.

mroeschke replied:
Agreed. I guess I was getting caught up on the semantics. I would just rather say `the ambiguous parameter dictates how ambiguous times should be handled.`

mroeschke commented on pandas/core/generic.py:
```diff
@@ -8623,6 +8623,13 @@ def tz_localize(self, tz, axis=0, level=None, copy=True,
         ambiguous : 'infer', bool-ndarray, 'NaT', default 'raise'
+            When clocks moved backward due to DST, ambiguous times may arise.
+            E.g. in Central European Time (UTC+01), when going from 03:00 DST
+            to 02:00 non-DST, the naive time 02:30:00 occurs both at
```
`the naive time 02:30:00` --> `02:30:00` local time

mroeschke commented:
Could you copy this explanation to the other occurrences of this argument? It needs to be added to `Timestamp`, `DatetimeIndex` and `NaT`

jorisvandenbossche commented:
@bartaelterman Thanks for the PR! 

In addition, what would also be very useful is an actual code example in an `Examples` section (feel free to add one in this PR, but also don't feel obliged, the PR is already an improvement)

bartaelterman commented:
Thanks for your feedback! I will wait for your reply on my comment above after which I will copy over the explanation to the other places. I'll look into the `Examples` section too.

bartaelterman committed:
Updating 'ambiguous'-param doc + update it on Timestamp, DatetimeIndex and NaT

This is following the discussion at
https://github.com/pandas-dev/pandas/pull/23408#discussion_r229109984

bartaelterman committed:
Add some examples for the ambiguous parameter for localizing timestamps

bartaelterman commented:
@mroeschke I've updated the doc as discussed and added it to `Timestamp`, `DatetimeIndex` and `NaT`.
@jorisvandenbossche I've added a couple of examples. However I'm not comfortable with the Cython stuff...
If you have more remarks, I am happy to modify it further.

bartaelterman commented:
So the docstring test for NaT have failed but I'm not sure what's wrong: 

```
=================================== FAILURES ===================================
_____________________________ test_NaT_docstrings ______________________________

    def test_NaT_docstrings():
        # GH#17327
        nat_names = dir(NaT)
    
        # NaT should have *most* of the Timestamp methods, with matching
        # docstrings.  The attributes that are not expected to be present in NaT
        # are private methods plus `ts_expected` below.
        ts_names = dir(Timestamp)
        ts_missing = [x for x in ts_names if x not in nat_names and
                      not x.startswith('_')]
        ts_missing.sort()
        ts_expected = ['freqstr', 'normalize',
                       'to_julian_date',
                       'to_period', 'tz']
        assert ts_missing == ts_expected
    
        ts_overlap = [x for x in nat_names if x in ts_names and
                      not x.startswith('_') and
                      callable(getattr(Timestamp, x))]
        for name in ts_overlap:
            tsdoc = getattr(Timestamp, name).__doc__
            natdoc = getattr(NaT, name).__doc__
>           assert tsdoc == natdoc
E           AssertionError: assert '\n        Co...ne.\n        ' == '\n        Con...ne.\n        '
E             Skipping 383 identical leading characters in diff, use -v to show
E             -           When clocks moved backward due to DST, ambiguous times may arise.
E             -             For example in Central European Time (UTC+01), when going from 03:00 DST
E             -             to 02:00 non-DST, 02:30:00 local time occurs both at 00:30:00 UTC
E             -             and at 01:30:00 UTC. In such a situation, the `ambiguous` parameter
E             -             dictates how ambiguous times should be handled.
E             - 
E             -             - bool contains flags to determine if time is dst or not (note
E             ? --
E             +           - bool contains flags to determine if time is dst or not (note
E                             that this flag is only applicable for ambiguous fall dst dates)
E                           - 'NaT' will return NaT for an ambiguous time
E                           - 'raise' will raise an AmbiguousTimeError for an ambiguous time
E               
E                       nonexistent : 'shift', 'NaT', default 'raise'
E                           A nonexistent time does not exist in a particular timezone
E                           where clocks moved forward due to DST.
E               
E                           - 'shift' will shift the nonexistent time forward to the closest
E                             existing time
E                           - 'NaT' will return NaT where there are nonexistent times
E                           - 'raise' will raise an NonExistentTimeError if there are
E                             nonexistent times
E               
E                           .. versionadded:: 0.24.0
E               
E                       errors : 'raise', 'coerce', default None
E                           - 'raise' will raise a NonExistentTimeError if a timestamp is not
E                              valid in the specified timezone (e.g. due to a transition from
E                              or to DST time). Use ``nonexistent='raise'`` instead.
E                           - 'coerce' will return NaT if the timestamp can not be converted
E                             into the specified timezone. Use ``nonexistent='NaT'`` instead.
E               
E                             .. deprecated:: 0.24.0
E               
E                       Returns
E                       -------
E                       localized : Timestamp
E               
E                       Raises
E                       ------
E                       TypeError
E                           If the Timestamp is tz-aware and tz is not None.

pandas/tests/scalar/test_nat.py:185: AssertionError
```

mroeschke commented on pandas/core/generic.py:
```diff
@@ -8650,6 +8656,50 @@ def tz_localize(self, tz, axis=0, level=None, copy=True,
+        Examples
+        --------
+
+        Localize local times:
```
For this set of examples, I think these methods should be showing examples for `Series.tz_localize` (the timeseries data is passed to `index=`) and not `Series.dt.tz_localize`(the timeseries data is pass to `data=` since that's where this docstring is linked

mroeschke commented on pandas/core/indexes/datetimes.py:
```diff
@@ -173,6 +179,45 @@ class DatetimeIndex(DatetimeArrayMixin, DatelikeOps, TimelikeOps,
     PeriodIndex : Index of Period data
     pandas.to_datetime : Convert argument to datetime
+
+    Examples
```
These example should live under the `tz_localize` method docstring: https://github.com/pandas-dev/pandas/blob/7191af9b47f6f57991abdb20625d9563877370a2/pandas/core/arrays/datetimes.py#L591

mroeschke commented on pandas/core/indexes/datetimes.py:
```diff
@@ -173,6 +179,45 @@ class DatetimeIndex(DatetimeArrayMixin, DatelikeOps, TimelikeOps,
+
+        Localize local times:
+
+        >>> s = pd.to_datetime(pd.Series(['2018-09-15 01:30:00']))
```
Don't need the Series call, the list will just return a DatetimeIndex.

mroeschke commented on pandas/core/indexes/datetimes.py:
```diff
@@ -173,6 +179,45 @@ class DatetimeIndex(DatetimeArrayMixin, DatelikeOps, TimelikeOps,
+        Localize local times:
+
+        >>> s = pd.to_datetime(pd.Series(['2018-09-15 01:30:00']))
+        >>> pd.DatetimeIndex(s, tz='CET')
```
Would be better to show `tz_localize` once this docstring is under the `tz_localize` method docstring.

bartaelterman committed:
Make docstring of NaT tz_localize equal to Timestamp tz_localize

bartaelterman committed:
show example for Series.tz_localize (not dt.tz_localize)

bartaelterman commented:
Should I add the docstring also to `core/arrays/datetimes.py`? And in that case, shall I add the example with `dt.tz_localize()` there? It seems to me that would end up [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.dt.tz_localize.html#pandas.Series.dt.tz_localize) which seems appropriate.

mroeschke commented:
Yes that'd be great @bartaelterman.

The examples look good.  Just looks like there's a doctest failure and pep8 issues.

bartaelterman committed:
add ambiguous docstring to arrays/datetimes

bartaelterman committed:
fix bug in the docstring

bartaelterman commented:
The Travis build is still failing, but I'm not sure why. All tests seem to pass and I don't see an error message, except at the very end:
`The command "ci/code_checks.sh" exited with 2.`
Any ideas what's going wrong? Thanks

TomAugspurger commented:
see https://travis-ci.org/pandas-dev/pandas/jobs/448947238#L2763

Some lines are too long.

bartaelterman committed:
chop off long lines in docstring

bartaelterman committed:
chop long lines in docstring

bartaelterman committed:
chop long line in docstring

TomAugspurger commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -835,7 +835,20 @@ def tz_localize_to_utc(ndarray[int64_t] vals, object tz, object ambiguous=None,
+          non-DST time (note that this flag is only applicable for ambiguous
+          times, but the array must have the same length as vals)
+        - bool if True, treat all vals as DST. If False, treat them as non-DST
+        - 'NaT' will return NaT where there are ambiguous times
```
Is this the string 'NaT', or the singleton pd.NaT (or either?) Can we clarify that?

TomAugspurger replied:
It seems like the allowed types are a bit inconsistent among all these functions. Can you check whether that's the docs being wrong? or are the just different?

mroeschke replied:
This will return the pd.NaT singleton.

TomAugspurger replied:
I was wondering about whether the input should be `pd.NaT` or `'NaT'`.

mroeschke replied:
Oh sorry, ambiguous accepts the input `'NaT` and should raise on `pd.NaT`'.

datapythonista reviewed (approved):
lgtm

But we should check whether the `ambiguos` accepts different values ni each function, or it's the documentation that doesn't document all cases. But it can be in a separate PR I think.

TomAugspurger commented:
Agreed. This is a useful change as is.
